include/flang/Lower/PFTBuilder.h

```diff
@@ -115,54 +116,81 @@ static constexpr bool isFunctionLike{common::HasMember<
-
-/// Function-like units contain lists of evaluations.  These can be simple
-/// statements or constructs, where a construct contains its own evaluations.
-struct Evaluation {
-  using EvaluationTuple = common::CombineTuples<ActionStmts, OtherStmts,
-                                                ConstructStmts, Constructs>;
-
-  /// Hide non-nullable pointers to the parse-tree node.
-  /// Build type std::variant<const A* const, const B* const, ...>
-  /// from EvaluationTuple type (std::tuple<A, B, ...>).
-  template <typename A>
-  using MakeRefType = const A *const;
-  using EvaluationVariant = common::CombineVariants<
-      common::MapTemplate<MakeRefType, EvaluationTuple>>;
-  template <typename A>
-  constexpr auto visit(A visitor) const {
-    return std::visit(common::visitors{
-                          [&](const auto *p) { return visitor(*p); },
-                          [&](auto &r) { return visitor(r); },
-                      },
-                      u);
+using SymbolLabelMap = llvm::DenseMap<const semantics::Symbol *, LabelSet>;
+
+/// Provide a variant like container that can holds constant references.
+/// It is used in the other classes to provide union of const references
+/// to parse-tree nodes.
+template <typename... A>
+class ReferenceVariant {
+public:
+  template <typename B>
+  using ConstRef = common::Reference<const B>;
+
+  ReferenceVariant() = delete;
+  template <typename B>
+  ReferenceVariant(const B &b) : u{ConstRef<B>{b}} {};
+
+  template <typename B>
+  constexpr const B &get() const {
+    return std::get<ConstRef<B>> > (u).get();
   }
-  template <typename A>
-  constexpr const A *getIf() const {
-    auto *ptr = std::get_if<MakeRefType<A>>(&u);
-    return ptr ? *ptr : nullptr;
+  template <typename B>
+  constexpr const B *getIf() const {
+    auto *ptr = std::get_if<ConstRef<B>>(&u);
+    return ptr ? &ptr->get() : nullptr;
   }
-  template <typename A>
+  template <typename B>
   constexpr bool isA() const {
-    return std::holds_alternative<MakeRefType<A>>(u);
+    return std::holds_alternative<ConstRef<B>>(u);
+  }
+  template <typename VISITOR>
+  constexpr auto visit(VISITOR &&visitor) const {
+    return std::visit(
+        common::visitors{[&visitor](auto ref) { return visitor(ref.get()); }},
+        u);
   }
 
-  Evaluation() = delete;
-  Evaluation(const Evaluation &) = delete;
-  Evaluation(Evaluation &&) = default;
+private:
+  std::variant<ConstRef<A>...> u;
+};
+
+template <typename A>
+struct MakeReferenceVariantHelper {};
+template <typename... A>
+struct MakeReferenceVariantHelper<std::variant<A...>> {
+  using type = ReferenceVariant<A...>;
+};
+template <typename... A>
+struct MakeReferenceVariantHelper<std::tuple<A...>> {
+  using type = ReferenceVariant<A...>;
+};
+template <typename A>
+using MakeReferenceVariant = typename MakeReferenceVariantHelper<A>::type;
+
+using EvaluationTuple =
+    common::CombineTuples<ActionStmts, OtherStmts, ConstructStmts, Constructs>;
```

Comment:
Should we really assume `Fortran` namespace anywhere in the bridge? Since we have a multitude of namespaces that we positively must deal with here, I'm not sure it is the best in terms of future-proofing to assume anything about the namespace defaulting one way or another.

Reply:
I saw you are not assuming it in Bridge.cpp. In `PFTBuilder.h`, I think it would make the tuple of parser statements types unreadable to add this constraint here.

Reply:
Ah. I forget this file is very close to the front-end.